### PR TITLE
[deps] Update RubyGems to 3.2 and Bundler 2.2

### PR DESCRIPTION
**Pull request by kares**
UPDATE: this PR removes tracked rubygems sources (except JRuby defaults) and also updates Bundler to 2.2
(this is aligned with `update_rubygems` as RGs 3.2 during update also installs latest Bundler)

### CODE REVIEW

**Comment by headius** [comment]
It would be good to get an update in and baking before we start prepping for 9.3 release.

**Comment by kares** [comment]
:+1: initially just wanted to check a test run, turns out there's one regression ([due](https://github.com/jruby/jruby/pull/6533/files#diff-1786bb3128a46cae1bf22dcae2d01ed4d39232d2632e04c54a327f3590ba8662R472)) JRuby is guarding with a spec:
```ruby
  it "should not create gem subdirectories on a non-file: URL" do
    Gem.ensure_gem_subdirectories("classpath:/bogus/classpath")
    File.exist?("classpath:/bogus/classpath").should be_false
    Gem.ensure_gem_subdirectories("file:")
    File.exist?("file:").should be_false
    Gem.ensure_gem_subdirectories("uri:file://bogus/classpath")
    File.exist?("uri:file:///nothing").should be_false
    File.exist?("uri:file://bogus/classpath").should be_false
  end
```
```
Failures:

  1) RubyGems extensions should not create gem subdirectories on a non-file: URL

     Failure/Error: Dir.mkdir path

     Errno::ENOTDIR: Not a directory - classpath:/bogus

     # org/jruby/RubyDir.java:632:in `mkdir'
     # ./lib/ruby/stdlib/fileutils.rb:235:in `fu_mkdir'
     # ./lib/ruby/stdlib/fileutils.rb:213:in `block in mkdir_p'
     # org/jruby/RubyArray.java:1891:in `reverse_each'
     # ./lib/ruby/stdlib/fileutils.rb:211:in `block in mkdir_p'
     # org/jruby/RubyArray.java:1809:in `each'
     # ./lib/ruby/stdlib/fileutils.rb:196:in `mkdir_p'
     # ./lib/ruby/stdlib/rubygems.rb:471:in `block in ensure_subdirectories'
```
spec originated all the way back to: https://github.com/jruby/jruby/commit/83dbc6313e9305371eb59d8aaab2cb0eb9b42101
.. believe this is there to handle embed cases where RGs would attempt to create it's gem layout
resolution submited upstream as: https://github.com/rubygems/rubygems/pull/4286

**Comment by kares** [comment]
seems to be also reproducing a AOT compilation issue (incorrect access modifier?) on the `test:jruby` suite:
```
Invalid gemspec in [/home/travis/build/jruby/jruby/lib/ruby/gems/shared/specifications/default/json-2.2.0-java.gemspec]: private method `name=' called for #<Gem::Specification:0x655a01d8>

NoMethodError: private method `satisfied_by?' called for #<Gem::Requirement:0x4d499d65>
  matching_specs at /home/travis/build/jruby/jruby/lib/ruby/stdlib/rubygems/dependency.rb:277
            each at org/jruby/RubyArray.java:1809
        find_all at org/jruby/RubyEnumerable.java:842
  matching_specs at /home/travis/build/jruby/jruby/lib/ruby/stdlib/rubygems/dependency.rb:276
        to_specs at /home/travis/build/jruby/jruby/lib/ruby/stdlib/rubygems/dependency.rb:299
         to_spec at /home/travis/build/jruby/jruby/lib/ruby/stdlib/rubygems/dependency.rb:319
             gem at /home/travis/build/jruby/jruby/lib/ruby/stdlib/rubygems/core_ext/kernel_gem.rb:62
          <main> at uri:classloader:/jruby/kernel/gem_prelude.rb:8
            load at org/jruby/RubyKernel.java:1009
          <main> at file:/home/travis/build/jruby/jruby/lib/jruby.jar!/jruby/preludes.rb:4
```

**Comment by deivid-rodriguez** [comment]
We sometimes get this error in CI: https://github.com/rubygems/rubygems/pull/3398/checks?check_run_id=1825431461. Could be a similar issue?

**Comment by kares** [comment]
@deivid-rodriguez hmm, seems different - while this seemed like a JIT modifier the RGs test case ends up with:
```
 ERROR:  While executing gem ... (NoMethodError)
    undefined method `ignored=' for nil:NilClass
	/home/runner/.rubies/jruby-9.2.14.0/lib/ruby/stdlib/rubygems/stub_specification.rb:193:in `to_spec'
	org/jruby/RubyArray.java:2577:in `map'
	/home/runner/.rubies/jruby-9.2.14.0/lib/ruby/stdlib/rubygems/specification.rb:748:in `_all'
	/home/runner/.rubies/jruby-9.2.14.0/lib/ruby/stdlib/rubygems/specification.rb:949:in `each'
	org/jruby/RubyEnumerable.java:1872:in `any?'
```
could be (JIT) related thought - definitely something worth investigating.

**Comment by headius** [comment]
The AOT issue appears to be how AOT scripts are loaded; we are not pushing a new frame for them. In the `satisfied_by?` case an autoload of `Gem::Version` fires in the middle of loading requirement.rb and properly sets its frame visibility to private. However it is the still-active frame from the body of `Gem::Requirement` and upon return to that body the rest of the methods get set private. Something similar is likely causing the gemspec errors.

I would not expect this to affect anyone running JRuby normally. The only script that is AOT compiled under normal circumstances is the main script.

**Comment by headius** [comment]
Note that master sets up and loads AOT code differently due to my work on GraalVM Native Image compilation, so it probably does not have the same issue. Merging to master should ignore the `Compiler` class and the change made for 9.2.

**Comment by headius** [comment]
Oh sorry, didn't realize I was pushing to your fork. 🙇‍♂️

**Comment by deivid-rodriguez** [comment]
Is it now fixed? That'd be awesome!

I understand that this doesn't affect "normal" usages of jruby, but it does affect our real usage of jruby in our tests, and it's been getting in the middle for a while. If this is fixed, we would no longer have any jruby related flaky failures :tada:.

**Comment by deivid-rodriguez** [comment]
Funny how I just got a different flaky failure 2 minutes after posting my previous message :rofl:. I just got this: https://github.com/rubygems/rubygems/pull/4371/checks?check_run_id=1893310735.

```
TEST_ENV_NUMBER=1 bin/rspec --tag=jruby --seed 40635 spec/bundler/bundler_spec.rb spec/bundler/definition_spec.rb spec/bundler/dep_proxy_spec.rb spec/bundler/dsl_spec.rb spec/bundler/environment_preserver_spec.rb spec/bundler/fetcher/compact_index_spec.rb spec/bundler/fetcher/downloader_spec.rb spec/bundler/fetcher_spec.rb spec/bundler/gem_version_promoter_spec.rb spec/bundler/index_spec.rb spec/bundler/installer/gem_installer_spec.rb spec/bundler/installer/parallel_installer_spec.rb spec/bundler/plugin/api/source_spec.rb spec/bundler/plugin/dsl_spec.rb spec/bundler/plugin/events_spec.rb spec/bundler/plugin_spec.rb spec/bundler/ruby_dsl_spec.rb spec/bundler/ruby_version_spec.rb spec/bundler/settings_spec.rb spec/bundler/source/git_spec.rb spec/bundler/source/rubygems/remote_spec.rb spec/bundler/source/rubygems_spec.rb spec/bundler/source_list_spec.rb spec/bundler/spec_set_spec.rb spec/bundler/uri_credentials_filter_spec.rb spec/bundler/vendored_persistent_spec.rb spec/bundler/yaml_serializer_spec.rb spec/cache/cache_path_spec.rb spec/cache/gems_spec.rb spec/cache/platform_spec.rb spec/commands/cache_spec.rb spec/commands/config_spec.rb spec/commands/doctor_spec.rb spec/commands/exec_spec.rb spec/commands/fund_spec.rb spec/commands/help_spec.rb spec/commands/info_spec.rb spec/commands/issue_spec.rb spec/commands/licenses_spec.rb spec/commands/list_spec.rb spec/commands/outdated_spec.rb spec/commands/show_spec.rb spec/commands/viz_spec.rb spec/install/allow_offline_install_spec.rb spec/install/deploy_spec.rb spec/install/failure_spec.rb spec/install/gemfile/gemspec_spec.rb spec/install/gemfile/git_spec.rb spec/install/gemfile/groups_spec.rb spec/install/gemfile/install_if_spec.rb spec/install/gemfile/lockfile_spec.rb spec/install/gemfile/platform_spec.rb spec/install/gemfile/sources_spec.rb spec/install/gemfile/specific_platform_spec.rb spec/install/gemfile_spec.rb spec/install/gems/env_spec.rb spec/install/gems/flex_spec.rb spec/install/gems/post_install_spec.rb spec/install/gems/resolving_spec.rb spec/install/gems/sudo_spec.rb spec/install/gemspecs_spec.rb spec/install/git_spec.rb spec/install/global_cache_spec.rb spec/install/path_spec.rb spec/install/security_policy_spec.rb spec/install/yanked_spec.rb spec/lock/lockfile_spec.rb spec/other/major_deprecation_spec.rb spec/other/platform_spec.rb spec/plugins/install_spec.rb spec/plugins/list_spec.rb spec/plugins/uninstall_spec.rb spec/quality_es_spec.rb spec/realworld/dependency_api_spec.rb spec/realworld/gemfile_source_header_spec.rb spec/realworld/mirror_probe_spec.rb spec/realworld/slow_perf_spec.rb spec/runtime/executable_spec.rb spec/runtime/gem_tasks_spec.rb spec/runtime/load_spec.rb spec/runtime/platform_spec.rb spec/update/gemfile_spec.rb spec/update/git_spec.rb spec/update/path_spec.rb
TEST_ENV_NUMBER=2 bin/rspec --tag=jruby --seed 54717 spec/bundler/build_metadata_spec.rb spec/bundler/cli_spec.rb spec/bundler/compact_index_client/updater_spec.rb spec/bundler/endpoint_specification_spec.rb spec/bundler/env_spec.rb spec/bundler/fetcher/base_spec.rb spec/bundler/fetcher/dependency_spec.rb spec/bundler/fetcher/index_spec.rb spec/bundler/friendly_errors_spec.rb spec/bundler/gem_helper_spec.rb spec/bundler/installer/spec_installation_spec.rb spec/bundler/lockfile_parser_spec.rb spec/bundler/mirror_spec.rb spec/bundler/plugin/api_spec.rb spec/bundler/plugin/index_spec.rb spec/bundler/plugin/installer_spec.rb spec/bundler/plugin/source_list_spec.rb spec/bundler/psyched_yaml_spec.rb spec/bundler/remote_specification_spec.rb spec/bundler/retry_spec.rb spec/bundler/rubygems_integration_spec.rb spec/bundler/settings/validator_spec.rb spec/bundler/shared_helpers_spec.rb spec/bundler/source/git/git_proxy_spec.rb spec/bundler/source/path_spec.rb spec/bundler/source_spec.rb spec/bundler/stub_specification_spec.rb spec/bundler/ui/shell_spec.rb spec/bundler/ui_spec.rb spec/bundler/version_ranges_spec.rb spec/bundler/worker_spec.rb spec/cache/git_spec.rb spec/cache/path_spec.rb spec/commands/add_spec.rb spec/commands/binstubs_spec.rb spec/commands/check_spec.rb spec/commands/clean_spec.rb spec/commands/console_spec.rb spec/commands/init_spec.rb spec/commands/inject_spec.rb spec/commands/install_spec.rb spec/commands/lock_spec.rb spec/commands/newgem_spec.rb spec/commands/open_spec.rb spec/commands/post_bundle_message_spec.rb spec/commands/pristine_spec.rb spec/commands/remove_spec.rb spec/commands/update_spec.rb spec/commands/version_spec.rb spec/install/binstubs_spec.rb spec/install/bundler_spec.rb spec/install/gemfile/eval_gemfile_spec.rb spec/install/gemfile/path_spec.rb spec/install/gemfile/ruby_spec.rb spec/install/gems/compact_index_spec.rb spec/install/gems/dependency_api_spec.rb spec/install/gems/fund_spec.rb spec/install/gems/mirror_spec.rb spec/install/gems/native_extensions_spec.rb spec/install/gems/standalone_spec.rb spec/install/gems/win32_spec.rb spec/install/prereleases_spec.rb spec/install/process_lock_spec.rb spec/install/redownload_spec.rb spec/lock/git_spec.rb spec/other/cli_dispatch_spec.rb spec/other/ext_spec.rb spec/plugins/command_spec.rb spec/plugins/hook_spec.rb spec/plugins/source/example_spec.rb spec/plugins/source_spec.rb spec/quality_spec.rb spec/realworld/double_check_spec.rb spec/realworld/edgecases_spec.rb spec/realworld/parallel_spec.rb spec/resolver/basic_spec.rb spec/resolver/platform_spec.rb spec/runtime/inline_spec.rb spec/runtime/require_spec.rb spec/runtime/setup_spec.rb spec/runtime/with_unbundled_env_spec.rb spec/update/gems/fund_spec.rb spec/update/gems/post_install_spec.rb spec/update/redownload_spec.rb
warning: thread "Ruby-0-Thread-382: /home/runner/work/rubygems/rubygems/bundler/spec/support/builders.rb:320" terminated with exception (report_on_exception is true):
Errno::EAGAIN: Resource temporarily unavailable - Resource temporarily unavailable
  waitpid at org/jruby/RubyProcess.java:919
..................F............

Failures:

  1) [TEST_ENV_NUMBER=1] bundle platform bundle update fails when ruby engine version doesn't match
     Failure/Error: Unable to find org/jruby/RubyProcess.java to read failed line

     Errno::EAGAIN:


       Commands:
       $ /home/runner/.rubies/jruby-9.2.14.0/bin/jruby -I/home/runner/work/rubygems/rubygems/bundler/spec -r/home/runner/work/rubygems/rubygems/bundler/spec/support/artifice/fail.rb -r/home/runner/work/rubygems/rubygems/bundler/spec/support/hax.rb /home/runner/work/rubygems/rubygems/bundler/tmp/1/gems/system/bin/bundle install
       Fetching source index from file:///home/runner/work/rubygems/rubygems/bundler/tmp/1/gems/remote2/
       Resolving dependencies...
       Using bundler 2.3.0.dev
       Fetching activesupport 2.3.5
       Fetching rack 1.0.0
       Installing rack 1.0.0
       Installing activesupport 2.3.5
       Fetching rack-obama 1.0
       Installing rack-obama 1.0
       Bundle complete! 2 Gemfile dependencies, 4 gems now installed.
       Use `bundle info [gemname]` to see where a bundled gem is installed.
       Post-install message from rack:
       Rack's post install message
       Post-install message from rack-obama:
       Rack-obama's post install message
       # $? => 0
       Resource temporarily unavailable - Resource temporarily unavailable
     # org/jruby/RubyProcess.java:919:in `waitpid'
     # 
     #   Showing full backtrace because every line was filtered out.
     #   See docs for RSpec::Configuration#backtrace_exclusion_patterns and
     #   RSpec::Configuration#backtrace_inclusion_patterns for more information.

Finished in 15 minutes 45 seconds (files took 0 seconds to load)
31 examples, 1 failure

Failed examples:

rspec ./spec/other/platform_spec.rb:562 # [TEST_ENV_NUMBER=1] bundle platform bundle update fails when ruby engine version doesn't match
```

**Comment by deivid-rodriguez** [comment]
I rerunning CI now to confirm it is a flaky but I'm pretty sure it is because I tend to recall it has happened previously.

**Comment by headius** [comment]
@deivid-rodriguez That EAGAIN in waitpid has been fixed on master and backported to 9.2, but that change has not been merged into this branch.

See #6477

**Comment by kares** [comment]
re-targeted and rebased off **master** ... let's see how it goes
(original intent was to see if there's any issues updating RGs on 9.2)

**Comment by kares** [comment]
I see RGs 3.2.6 was pushed since (no need for https://github.com/jruby/jruby/pull/6533/commits/55a76b3bd37bf641b6b04107fe8a81f432cacb53) 
\+ a few 3.2 releases recently latest being 3.2.10, will update once I read the changelogs.

**Comment by kares** [comment]
looking good, only 2 (seemingly unrelated) failures from *test:jruby* suite:
```
     194: 
     195:     def test_tcp_socket_get_keep_cnt
     196:       socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
  => 197:       assert_instance_of(Integer, socket.getsockopt(Socket::SOL_TCP, Socket::TCP_KEEPCNT).int)
     198:     ensure
     199:       socket.close
     200:     end

/home/travis/build/jruby/jruby/test/jruby/test_socket.rb:197:in `test_tcp_socket_get_keep_cnt'
org/jruby/ext/socket/Option.java:208:in `int'

Error: test_tcp_socket_get_keep_cnt(SocketTest): TypeError: size differ.  expected as sizeof(int)=4 but 2

================================================================================

E

================================================================================

     180:   if RbConfig::CONFIG['target_os'] == 'linux'
     181:     def test_tcp_socket_get_keep_idle
     182:       socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
  => 183:       assert_instance_of(Integer, socket.getsockopt(Socket::SOL_TCP, Socket::TCP_KEEPIDLE).int)
     184:     ensure
     185:       socket.close
     186:     end

/home/travis/build/jruby/jruby/test/jruby/test_socket.rb:183:in `test_tcp_socket_get_keep_idle'
org/jruby/ext/socket/Option.java:208:in `int'

Error: test_tcp_socket_get_keep_idle(SocketTest): TypeError: size differ.  expected as sizeof(int)=4 but 0
```

**Comment by deivid-rodriguez** [comment]
It's 3.2.11 now, actually :)

**Comment by deivid-rodriguez** [comment]
> That EAGAIN in waitpid has been fixed on master and backported to 9.2, but that change has not been merged into this branch.

Great, thanks @headius! That means once the fix is released with the next 9.2.x release and we upgrade our CI, we should be flaky free :)

**Comment by kares** [comment]
and we're all :green_heart: now with latest & greatest RGs 3.2.11, thanks again David for handling the updates JRuby needed!

**Comment by headius** [comment]
> we should be flaky free :)

Here's hoping! 😅

**Comment by headius** [comment]
Should we consider updating 9.2 as well @kares @enebo? RVM at least updates RG anyway, and folks like @deivid-rodriguez have been cranking out bug fixes and updates rapidly over the past year.

**Comment by kares** [comment]
:+1: happy to re-target back to 9.2 if decided ... CA certificates were updated in RGs 3.2 which might prolong 9.2's ttl
it's still going to need that AOT fix, right?

**Comment by headius** [comment]
> it's still going to need that AOT fix, right?

Yeah I will backport that. It was small and low-risk.

> the other issue is RGs 3.2 brings Bundler along

Upgrade RG but not ship bundler perhaps, as we have been doing for 9.2?

If we want to start shipping bundler with 9.2 we will also need to set it as a default gem and pull in some other default gem fixes (for bin scripts etc) that are only on master.

**Comment by kares** [comment]
> Upgrade RG but not ship bundler perhaps, as we have been doing for 9.2?

We did not need to exclude anything for current RGs shipped with 9.2 - no Bundler packed.
I am fine with excluding - hopefully won't confuse users that they have RGs 3.2 wout Bundler 
(there will be no way for them to system upgrade to a 3.2 version wout pulling in Bundler).

**Comment by headius** [comment]
> We did not need to exclude anything for current RGs shipped with 9.2 - no Bundler packed.

Ahh, that is a good point.

It is pretty annoying that the RubyGems system update shoves bundler into the stdlib rather than just installing it as a gem. If we don't set it up as a default gem on JRuby 9.2, then there's no way to update bundler independently after you have installed it into stdlib (it won't see the gem).

Maybe we should just leave well enough alone on 9.2 after all.

**Comment by kares** [comment]
> Maybe we should just leave well enough alone on 9.2 after all.

Probably not worth the hustle for the last-ish 9.2 release, in which case this PR targeting master should be good to :ship:.
We still end up w Bundler in stdlib but I believe it's expected with RGs 3.2.

**Comment by headius** [comment]
> We still end up w Bundler in stdlib but I believe it's expected with RGs 3.2.

Unfortunately there is a wrinkle with this: Bundler needs to be installed as a default gem. This PR currently does not install it as a default gem, which means it can't be upgraded.

See lib/pom.rb on master; rather than updating bundler files with this PR we should update that pom to install bundler as a default gem.

@deivid-rodriguez Sorry to keep bugging you for RubyGems/Bundler issues, but it is rather inconvenient that the RubyGems upgrade installs the bundler files directly in stdlib, without making Bundler a default gem. Can we do something to improve this?

File: lib/pom.rb [hunk 1]
```diff
@@ -15,7 +15,7 @@ def log(message=nil)
 end
 
 default_gems = [
-    ['bundler', '1.17.3'],
+    # Bundler is shipped with RubyGems
```
Comment (headius) [inline lib/pom.rb:18]: We need to keep Bundler installing as a default gem or it will not be upgradable by users.

**Review by headius — CHANGES_REQUESTED** [review]
Need to remove bundler files and have them installed as part of the build. Bundler must be a default gem in order for it to be upgradable by users.

**Comment by deivid-rodriguez** [comment]
> @deivid-rodriguez Sorry to keep bugging you for RubyGems/Bundler issues, but it is rather inconvenient that the RubyGems upgrade installs the bundler files directly in stdlib, without making Bundler a default gem. Can we do something to improve this?

Rubygems should install bundler as a default gem, can you clarify the question?

**Comment by deivid-rodriguez** [comment]
If bundler is not getting installed as a default gem by `gem update --system`, that's unexpected.

**Comment by headius** [comment]
@deivid-rodriguez Confirmed you are right, Bundler does get installed as a default gem:

```
[] ~/projects/jruby-9.2 $ gem list bundler

*** LOCAL GEMS ***



[] ~/projects/jruby-9.2 $ gem update --system --no-document
Updating rubygems-update
Successfully installed rubygems-update-3.2.11
Installing RubyGems 3.2.11
  Successfully built RubyGem
  Name: bundler
  Version: 2.2.11
  File: bundler-2.2.11.gem
Bundler 2.2.11 installed
RubyGems 3.2.11 installed
Regenerating binstubs
Regenerating plugins
...

[] ~/projects/jruby-9.2 $ gem list bundler

*** LOCAL GEMS ***

bundler (default: 2.2.11)
```

There is a separate bug with how it generates the `gem` binscript, I will file an issue for that.

**Comment by headius** [comment]
@kares To clarify, it does appear that gem update installs bundler as a default gem. However I don't think we want to version these files. I can see a couple options:

* Per my review comments, revert to having bundler installed as a default gem by our build and don't version the files. This is perhaps less convenient since RubyGems will try to install the files again next time we update it.
* Remove both Bundler and RubyGems from our repository and run the RubyGems install as part of our build. This would avoid us versioning those files and should do the right thing for local builds and release dists.

**Comment by headius** [comment]
gem update --system breaks gem binscript on JRuby: https://github.com/rubygems/rubygems/issues/4404

**Comment by kares** [comment]
Have removed the rubygems files from stdlib (except the JRuby rubygems/defaults).
Setup to install rubygems-update while skipping copy of the bin executables to JRuby's bin folder.

**Review by headius — APPROVED** [review]
Looks great! Love to see deleted files!

**Pull request merged**